### ISSUE TITLE
prod / aot error for demo (likely a `lazyModules` problem)

### ISSUE THREAD

**Issue by Meligy** [apoterenko/ngx-dynamic-template#44]
Hello again,
Sorry for bothering many times. I just thought I'd let you know about this one.

If you run the demo from this repository with AoT on, it compiles with no errors, but then you get an error in the browser dev toolbar console.

## Steps To Reproduce

Just go to the demo folder and run:

```
npm start -- -prod --aot
```
(`npm start` is just `ng serve` -- also probably `--aot` is redundant here, but just to explain)

## Errors

After a successful compilation, if you open the browser to http://localhost:4200, you get

1. Secnario 2 does not load
2. You get the following browser console errors:

```
vendor.55c5451….bundle.js:148 ERROR Error: Uncaught (in promise): Error: Cannot find module 'app/lazy/lazy.module'.
Error: Cannot find module 'app/lazy/lazy.module'.
    at r (main.943b1f5….bundle.js:1)
    at t.loadAndCompile (vendor.55c5451….bundle.js:330)
    at t.load (vendor.55c5451….bundle.js:330)
    at e.t.buildModule (0.212b13c….chunk.js:1)
    at e.t.ngOnChanges (0.212b13c….chunk.js:1)
    at sn (vendor.55c5451….bundle.js:232)
    at Zn (vendor.55c5451….bundle.js:260)
    at zn (vendor.55c5451….bundle.js:260)
    at fr (vendor.55c5451….bundle.js:260)
    at Object.updateDirectives (0.212b13c….chunk.js:148)
    at r (main.943b1f5….bundle.js:1)
    at t.loadAndCompile (vendor.55c5451….bundle.js:330)
    at t.load (vendor.55c5451….bundle.js:330)
    at e.t.buildModule (0.212b13c….chunk.js:1)
    at e.t.ngOnChanges (0.212b13c….chunk.js:1)
    at sn (vendor.55c5451….bundle.js:232)
    at Zn (vendor.55c5451….bundle.js:260)
    at zn (vendor.55c5451….bundle.js:260)
    at fr (vendor.55c5451….bundle.js:260)
    at Object.updateDirectives (0.212b13c….chunk.js:148)
    at u (polyfills.72d7bf5….bundle.js:43)
    at u (polyfills.72d7bf5….bundle.js:43)
    at polyfills.72d7bf5….bundle.js:43
    at t.invokeTask (polyfills.72d7bf5….bundle.js:36)
    at Object.onInvokeTask (vendor.55c5451….bundle.js:316)
    at t.invokeTask (polyfills.72d7bf5….bundle.js:36)
    at r.runTask (polyfills.72d7bf5….bundle.js:36)
    at o (polyfills.72d7bf5….bundle.js:36)
```

```
ERROR Error: Uncaught (in promise): Error: Cannot find module 'app/lazy2/lazy2.module'.
Error: Cannot find module 'app/lazy2/lazy2.module'.
    at r (main.943b1f5….bundle.js:1)
    at t.loadAndCompile (vendor.55c5451….bundle.js:330)
    at t.load (vendor.55c5451….bundle.js:330)
    at e.t.buildModule (0.212b13c….chunk.js:1)
    at e.t.ngOnChanges (0.212b13c….chunk.js:1)
    at sn (vendor.55c5451….bundle.js:232)
    at Zn (vendor.55c5451….bundle.js:260)
    at zn (vendor.55c5451….bundle.js:260)
    at fr (vendor.55c5451….bundle.js:260)
    at Object.View_t_0.co [as updateDirectives] (t.ngfactory.js:71)
    at r (main.943b1f5….bundle.js:1)
    at t.loadAndCompile (vendor.55c5451….bundle.js:330)
    at t.load (vendor.55c5451….bundle.js:330)
    at e.t.buildModule (0.212b13c….chunk.js:1)
    at e.t.ngOnChanges (0.212b13c….chunk.js:1)
    at sn (vendor.55c5451….bundle.js:232)
    at Zn (vendor.55c5451….bundle.js:260)
    at zn (vendor.55c5451….bundle.js:260)
    at fr (vendor.55c5451….bundle.js:260)
    at Object.View_t_0.co [as updateDirectives] (t.ngfactory.js:71)
    at u (polyfills.72d7bf5….bundle.js:43)
    at u (polyfills.72d7bf5….bundle.js:43)
    at polyfills.72d7bf5….bundle.js:43
    at t.invokeTask (polyfills.72d7bf5….bundle.js:36)
    at Object.onInvokeTask (vendor.55c5451….bundle.js:316)
    at t.invokeTask (polyfills.72d7bf5….bundle.js:36)
    at r.runTask (polyfills.72d7bf5….bundle.js:36)
    at o (polyfills.72d7bf5….bundle.js:36)
```

![image](https://cloud.githubusercontent.com/assets/171637/26295411/1cd4e8c8-3f0d-11e7-9382-affd136bcd4f.png)

Since scenario 2 is the one that does not load, this is probably an issue in implementing #43.

**Comment by apoterenko** [comment]
Yeah, I know about this issue. I will try to fix it as soon as possible

**Comment by Meligy** [comment]
Just for the record, v2.1.6 seems to have broken both JIT and AoT (still only for scenario 2 / lazy modules).

### JIT Errors

```
core.es5.js:1084 ERROR Error: Uncaught (in promise): Error: Cannot find module 'app/lazy/lazy.module.ngfactory'.
Error: Cannot find module 'app/lazy/lazy.module.ngfactory'.
    at webpackAsyncContext (src async:17)
    at SystemJsNgModuleLoader.loadFactory (core.es5.js:5915)
    at DynamicDirective.webpackJsonp.271.DynamicBase.buildModule (dynamic.base.js:85)
    at DynamicDirective.webpackJsonp.271.DynamicBase.ngOnChanges (dynamic.base.js:28)
    at checkAndUpdateDirectiveInline (core.es5.js:10812)
    at checkAndUpdateNodeInline (core.es5.js:12238)
    at checkAndUpdateNode (core.es5.js:12177)
    at debugCheckAndUpdateNode (core.es5.js:12880)
    at debugCheckDirectivesFn (core.es5.js:12821)
    at Object.eval [as updateDirectives] (DynamicHolderComponent.html:11)
    at webpackAsyncContext (src async:17)
    at SystemJsNgModuleLoader.loadFactory (core.es5.js:5915)
    at DynamicDirective.webpackJsonp.271.DynamicBase.buildModule (dynamic.base.js:85)
    at DynamicDirective.webpackJsonp.271.DynamicBase.ngOnChanges (dynamic.base.js:28)
    at checkAndUpdateDirectiveInline (core.es5.js:10812)
    at checkAndUpdateNodeInline (core.es5.js:12238)
    at checkAndUpdateNode (core.es5.js:12177)
    at debugCheckAndUpdateNode (core.es5.js:12880)
    at debugCheckDirectivesFn (core.es5.js:12821)
    at Object.eval [as updateDirectives] (DynamicHolderComponent.html:11)
    at resolvePromise (zone.js:769)
    at resolvePromise (zone.js:740)
    at zone.js:817
    at ZoneDelegate.invokeTask (zone.js:424)
    at Object.onInvokeTask (core.es5.js:4140)
    at ZoneDelegate.invokeTask (zone.js:423)
    at Zone.runTask (zone.js:191)
    at drainMicroTaskQueue (zone.js:584)
    at <anonymous>
```

```
core.es5.js:1084 ERROR Error: Uncaught (in promise): Error: Cannot find module 'app/lazy2/lazy2.module.ngfactory'.
Error: Cannot find module 'app/lazy2/lazy2.module.ngfactory'.
    at webpackAsyncContext (src async:17)
    at SystemJsNgModuleLoader.loadFactory (core.es5.js:5915)
    at DynamicDirective.webpackJsonp.271.DynamicBase.buildModule (dynamic.base.js:85)
    at DynamicDirective.webpackJsonp.271.DynamicBase.ngOnChanges (dynamic.base.js:28)
    at checkAndUpdateDirectiveInline (core.es5.js:10812)
    at checkAndUpdateNodeInline (core.es5.js:12238)
    at checkAndUpdateNode (core.es5.js:12177)
    at debugCheckAndUpdateNode (core.es5.js:12880)
    at debugCheckDirectivesFn (core.es5.js:12821)
    at Object.View_dynamicComponentClass_0.co [as updateDirectives] (dynamicComponentClass.html:4)
    at webpackAsyncContext (src async:17)
    at SystemJsNgModuleLoader.loadFactory (core.es5.js:5915)
    at DynamicDirective.webpackJsonp.271.DynamicBase.buildModule (dynamic.base.js:85)
    at DynamicDirective.webpackJsonp.271.DynamicBase.ngOnChanges (dynamic.base.js:28)
    at checkAndUpdateDirectiveInline (core.es5.js:10812)
    at checkAndUpdateNodeInline (core.es5.js:12238)
    at checkAndUpdateNode (core.es5.js:12177)
    at debugCheckAndUpdateNode (core.es5.js:12880)
    at debugCheckDirectivesFn (core.es5.js:12821)
    at Object.View_dynamicComponentClass_0.co [as updateDirectives] (dynamicComponentClass.html:4)
    at resolvePromise (zone.js:769)
    at resolvePromise (zone.js:740)
    at zone.js:817
    at ZoneDelegate.invokeTask (zone.js:424)
    at Object.onInvokeTask (core.es5.js:4140)
    at ZoneDelegate.invokeTask (zone.js:423)
    at Zone.runTask (zone.js:191)
    at drainMicroTaskQueue (zone.js:584)
    at <anonymous>
```


### AoT / Prod Errors

```
vendor.55c54517c01c238db48e.bundle.js:148 ERROR Error: Uncaught (in promise): Error: Unexpected value 'n' imported by the module 't'. Please add a @NgModule annotation.
Error: Unexpected value 'n' imported by the module 't'. Please add a @NgModule annotation.
    at g (0.ba0a0f1a17b5e4a742b9.chunk.js:1)
    at 0.ba0a0f1a17b5e4a742b9.chunk.js:127
    at Array.forEach (<anonymous>)
    at t.Uk0N.t.getNgModuleMetadata (0.ba0a0f1a17b5e4a742b9.chunk.js:127)
    at t.Uk0N.t._loadModules (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t._compileModuleAndAllComponents (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t.compileModuleAndAllComponentsAsync (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at 0.ba0a0f1a17b5e4a742b9.chunk.js:1
    at t.invoke (polyfills.e455e12927face798d31.bundle.js:36)
    at Object.onInvoke (vendor.55c54517c01c238db48e.bundle.js:316)
    at g (0.ba0a0f1a17b5e4a742b9.chunk.js:1)
    at 0.ba0a0f1a17b5e4a742b9.chunk.js:127
    at Array.forEach (<anonymous>)
    at t.Uk0N.t.getNgModuleMetadata (0.ba0a0f1a17b5e4a742b9.chunk.js:127)
    at t.Uk0N.t._loadModules (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t._compileModuleAndAllComponents (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t.compileModuleAndAllComponentsAsync (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at 0.ba0a0f1a17b5e4a742b9.chunk.js:1
    at t.invoke (polyfills.e455e12927face798d31.bundle.js:36)
    at Object.onInvoke (vendor.55c54517c01c238db48e.bundle.js:316)
    at u (polyfills.e455e12927face798d31.bundle.js:43)
    at polyfills.e455e12927face798d31.bundle.js:43
    at t.invokeTask (polyfills.e455e12927face798d31.bundle.js:36)
    at Object.onInvokeTask (vendor.55c54517c01c238db48e.bundle.js:316)
    at t.invokeTask (polyfills.e455e12927face798d31.bundle.js:36)
    at r.runTask (polyfills.e455e12927face798d31.bundle.js:36)
    at o (polyfills.e455e12927face798d31.bundle.js:36)
    at <anonymous>
```

```
vendor.55c54517c01c238db48e.bundle.js:148 ERROR Error: Uncaught (in promise): Error: Unexpected value 't' imported by the module 't'. Please add a @NgModule annotation.
Error: Unexpected value 't' imported by the module 't'. Please add a @NgModule annotation.
    at g (0.ba0a0f1a17b5e4a742b9.chunk.js:1)
    at 0.ba0a0f1a17b5e4a742b9.chunk.js:127
    at Array.forEach (<anonymous>)
    at t.Uk0N.t.getNgModuleMetadata (0.ba0a0f1a17b5e4a742b9.chunk.js:127)
    at t.Uk0N.t._loadModules (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t._compileModuleAndAllComponents (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t.compileModuleAndAllComponentsAsync (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t.compileModuleAndAllComponentsAsync (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at 0.ba0a0f1a17b5e4a742b9.chunk.js:1
    at t.invoke (polyfills.e455e12927face798d31.bundle.js:36)
    at g (0.ba0a0f1a17b5e4a742b9.chunk.js:1)
    at 0.ba0a0f1a17b5e4a742b9.chunk.js:127
    at Array.forEach (<anonymous>)
    at t.Uk0N.t.getNgModuleMetadata (0.ba0a0f1a17b5e4a742b9.chunk.js:127)
    at t.Uk0N.t._loadModules (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t._compileModuleAndAllComponents (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t.compileModuleAndAllComponentsAsync (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at t.Uk0N.t.compileModuleAndAllComponentsAsync (0.ba0a0f1a17b5e4a742b9.chunk.js:148)
    at 0.ba0a0f1a17b5e4a742b9.chunk.js:1
    at t.invoke (polyfills.e455e12927face798d31.bundle.js:36)
    at u (polyfills.e455e12927face798d31.bundle.js:43)
    at polyfills.e455e12927face798d31.bundle.js:43
    at t.invokeTask (polyfills.e455e12927face798d31.bundle.js:36)
    at Object.onInvokeTask (vendor.55c54517c01c238db48e.bundle.js:316)
    at t.invokeTask (polyfills.e455e12927face798d31.bundle.js:36)
    at r.runTask (polyfills.e455e12927face798d31.bundle.js:36)
    at o (polyfills.e455e12927face798d31.bundle.js:36)
    at <anonymous>
```

It's horrible how Angular's AoT very different way of working from JIT makes this kind of stuff extremely hard :(

**Comment by apoterenko** [comment]
Yeah, I am still trying to understand why System.import inside SystemJsNgModuleLoader does not work at AoT mode

**Comment by Meligy** [comment]
If I just do `npm start -- --aot` (without `-prod`) it shows that it's talking about the correct module (`LazyModule` in first error, and `Lazy2Module` in second). So, it seems to be reaching the right object / class, but not sure why it thinks it doesn't have the `NgModule` decorator.

Maybe debugging in the browser with just `--aot` but not `-prod` can help.

**Comment by apoterenko** [comment]
Sorry, I'm at work just now. Please wait :)

**Comment by Meligy** [comment]
Oh. Yes, sure. Sorry, I thought you were debugging it, and was hoping I got help with the debugging myself.

I could only reach that `compileModuleAndAllComponentsAsync` at some point calls `getNgModuleMetadata` with an object that looks like this:

![image](https://cloud.githubusercontent.com/assets/171637/26311132/8a02eb82-3f46-11e7-9fc0-fb25d7b4fcac.png)

With a `decorators` property for `CommonModule`, but none for `LazyModule`.

But anyway, sorry for bothering you so much. ** it's obviously NOT urgent ** and even if it's, you are not paid to do it or anything and just doing this on your own.

I'm really sorry. I am just trying to see if I can debug it on my own and provide any help.

I will not update this unless I find something useful, or be able to push a PR or something.

Thanks again, and sorry for annoying you.

**Comment by apoterenko** [comment]
No problem)

Right now, I have two challenging problems. I need to think about how to solve them. For more details please see the comment

https://github.com/apoterenko/ngx-dynamic-template/commit/7e8c622cc4552949c1f7965bdd790a0e024624ee

**Comment by Meligy** [comment]
Yeah. I had the same problem when I tried to achieve this with `NgComponentTemplate` after giving up on debugging this issue.

Here's the approach that worked for me in case it gives any inspiration:

```html
<button (click)="loadDynamic()">Load Component</button>

<ng-container *ngIf="lazyComponent">
  <ng-container 
    *ngComponentOutlet="lazyComponent.component;ngModuleFactory: lazyComponent.moduleFactory">
  </ng-container>
</ng-container>
```

```typescript
export class AppComponent {
  lazyComponent: {
    component: Type<any>,
    moduleFactory: NgModuleFactory<any>
  } = null;

  constructor(private factoryLoader: NgModuleFactoryLoader) {
  }

  loadDynamic() {
    this.factoryLoader.load('./mod-two/mod-two.module#ModTwoModule')
      .then(factory => {

        this.lazyComponent = {
          moduleFactory: factory,
          component: factory.moduleType['bootstrap']
        };
      });
  }
}
```

That `bootstrap` thing is not the `@NgModule()` bootstrap property. I was hoping it would be, but like you I couldn't read the metadata (I tried via `Reflect` class, didn't know what else to try). So, I ended up setting it as a `static` property like:

```typescript
@NgModule({
  imports: [
    CommonModule, 
  ],
  declarations: [CatOneComponent],
  entryComponents: [CatOneComponent],
  bootstrap: [CatOneComponent]
})
export class ModTwoModule { 
  static bootstrap = CatOneComponent;
}
```

**Comment by apoterenko** [comment]
Yeah, we have no other choice as pass static meta info inside lazy loaded components and modules, and then build dynamic component and modules proxies recursively  :(

**Comment by apoterenko** [comment]
Sorry, too busy doing main jobs

> Error: Cannot find module

This issue has been fixed at 2.1.12 by true way. (https://github.com/apoterenko/ngx-dynamic-template/commit/a00a7e199bc2ffa6a042d642f85d376d919d8d73).

**Comment by apoterenko** [comment]
The solution via DynamicLazyMetadata works fine for me (2.1.14). See updated README

```
@DynamicLazyMetadata(moduleMetaData)
@NgModule(moduleMetaData)
export class LazyModule {
}
```

**Comment by apoterenko** [comment]
@Meligy FYI: Angular itself does not remove annotations. See https://github.com/angular/angular-cli/pull/4301. Maybe there is a configurable option like skipMetadata applied to angular-cli (tsconfig.ts maybe).

**Comment by apoterenko** [comment]
The issue has been resolved at 2.1.18 via Router.loadChildren which is being replaced with function by the cli-compiler and is being called by ngx-dynamic module. For more details, please see the main commit https://github.com/apoterenko/ngx-dynamic-template/commit/efce38777646eaaba4ac01a127c899c006328562
